Codechange: Sort Engine ID mapping for fast lookups

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

EngineOverrideManager records the grfid and localid of each engine, so that after save-load the correct engines are used again.

The looking up the EngineID, the complete list is traversed to find the entry by grfid and localid. As the list gets longer this takes more an more time.

An example of this Iron Horse which has approximately 10,000 engines.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Reorganise the list into 4 lists (one for each vehicle type) sorted by grfid and localid. This is built in a pre-sorted manner so that building the list itself can use a binary search instead of a linear scan. The overhead of moving the elements in the vector is tiny compared to performing a linear scan.

The original way of accessing the list with EngineID as the index is no longer possible, and now requires a linear scan to find it. Therefore as a prerequisite, some places that used to use the mapping now use the Engine's grfprops instead, which gains a grfid field.

These changes have no effect when no NewGRFs are loaded as in that case there no lookups during the game.

The performance impact depends on how often NewGRFs access information that needs the lookup, but in my testing there's an improvement of nearly 10,000x. With Iron Horse and several other vehicle NewGRFs loaded:

|       | Master | This PR |
| ----- | ------- | -------- |
| start game with iron horse | 865,156µs | 249µs | 
| per 100,000 calls | 637,596µs | 77µs |
| total over loading and 1 year | 19,765,466µs | 2,379µs |



<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

I didn't change the saveload format, so saving requires building a temporary list in EngineID order, and loading requires sorting the list into grfid/localid order (although the latter would always be needed for existing savegames.)

Loading now uses `_engine_mngr.SetID()` directly per entry, so it's automatically inserted into the correct order.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Codechange: Don't inherit EngineOverrideManager from std::vector.

Inheriting from std::vector means some operations are needlessly complex, and shouldn't really be done anyway.

PeterN committed:
Codechange: Sort EngineOverrideManager for fast lookups.

Allows quickly finding the EngineID given the type, grfid and local id of an engine, instead a linear scan.

This can reduce loading time when lots of engines are present and also affects performance in-game.
Lookup can be on the order of 10000 times faster.

PeterN committed:
Codechange: Store grfid with entity grfprops.

This allows using the grfid without having to dereference the grffile pointer.

Uses no extra storage as it fits within otherwise wasted padding space.

rubidium42 commented on src/engine.cpp:
```diff
@@ -508,14 +508,10 @@ bool Engine::IsVariantHidden(CompanyID c) const
  */
 void EngineOverrideManager::ResetToDefaultMapping()
 {
-	this->clear();
```
Second commit, commit message "some oeperations" -> "some operations"?

rubidium42 commented on src/engine.cpp:
```diff
@@ -508,14 +508,23 @@ bool Engine::IsVariantHidden(CompanyID c) const
 	}
 }
 
+struct EngineIDMappingComparator {
```
This has become a `Projection` instead of a `Comparator`. I am assuming you went from `std::lower_bound` to `std::ranges::lower_bound` at some point.

rubidium42 commented on src/newgrf.cpp:
```diff
@@ -653,8 +653,9 @@ static Engine *GetNewEngine(const GRFFile *file, VehicleType type, uint16_t inte
-			eid.grfid = scope_grfid; // Note: this is INVALID_GRFID if dynamic_engines is disabled, so no reservation
+			/* Replace an existing slot, so remove the existing map entry for it. */
+			_engine_mngr.EraseID(type, engine);
+			_engine_mngr.SetID(type, internal_id, scope_grfid, std::min<uint8_t>(internal_id, _engine_counts[type]), e->index);
```
This made me question whether this is suboptimal. Why wouldn't you just reuse the original location, but I reckon the primary reason this gets called is when it's replaced by an engine with a different grfid. Is it maybe worth to describe that?
'Replace an existing slot. Given the grfid (or internal_id) is likely different, the replacement will end up in a different location, so erase and add'

PeterN replied:
Yup. A rotate between the old and new position would allow the slot to be moved, moving only entries between them without moving everything after. But...

Maybe it's worth doing, it's just algorithms, and gives me an excuse to extract that rotate pair.

PeterN replied:
This can definitely be made more efficient, the split-functionality / abstraction between GetNewEngine and EngineOverrideManager makes it more work than it can be.

e.g. we've already got an iterator with the existing location, so the linear search in EraseID is pointless.

rubidium42 commented on src/newgrf_engine.cpp:
```diff
@@ -1311,17 +1311,17 @@ void AlterVehicleListOrder(EngineID engine, uint target)
 
 	/* 1. Sort by engine type */
-	if (id_a.type != id_b.type) return (int)id_a.type < (int)id_b.type;
+	if (engine_a.type != engine_b.type) return (int)engine_a.type < (int)engine_b.type;
```
Maybe a sprinkling of `static_cast` here and for the sorting of `local_id`?

rubidium42 commented on src/saveload/engine_sl.cpp:
```diff
@@ -189,8 +189,23 @@ struct EIDSChunkHandler : ChunkHandler {
 
+		/* Get total entries required for flat map. */
+		size_t total = 0;
+		for (const auto &mapping : _engine_mngr.mappings) { total += std::size(mapping); }
```
Doesn't coding style practically dictate using newlines when using curly brackets for a `for`? Same with the next loop.

PeterN committed:
Codechange: Add Slide container helper function.

This function will move the selected range between first and last to position, rotating elements as necessary.

Returns iterators to the new positions.

rubidium42 commented on src/core/container_func.hpp:
```diff
@@ -46,4 +46,19 @@ int find_index(Container const &container, typename Container::const_reference i
 
+/**
+ * Move elements between first and last to a new position, rotating elements in between as necessary.
+ * @param first Iterator to first lement to move.
```
```suggestion
 * @param first Iterator to first element to move.
```

michicc commented on src/core/container_func.hpp:
```diff
@@ -46,4 +46,19 @@ int find_index(Container const &container, typename Container::const_reference i
+ * @returns Iterators to first and last after being moved.
+ */
+template <typename TIter>
+auto Slide(TIter first, TIter last, TIter position) -> std::pair<TIter, TIter>
```
Is that kinda `std::shift_left, std::shift_right`, except with an interator instead of a distance?

PeterN replied:
Looks like it. In this case it would require iterator-arithmetic.

rubidium42 replied:
Is the point of this comment that distance might be considered as parameter instead of iterator? That seems to be more work, both in calculating the distance and in mapping the distance back reliably (if it were allowed to be negative, handling overflows, etc).

Or is the point that `std::rotate` can be replaced by `std::shift_left/std::right`? As I do not really see how shift and rotate are similar, as shift 'throws away' the bits shifted out where rotate does not.
